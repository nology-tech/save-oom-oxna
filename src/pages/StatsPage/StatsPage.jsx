import "./StatsPage.scss";
import AvatarHeading from "../../components/AvatarHeading/AvatarHeading";
import Graph from "../../components/Graph/Graph";
import DropdownButton from "../../components/Dropdown/Dropdown";

const StatsPage = () => {
  return (
    <div className="stats-page">
      <section className="stats-page__stats">
        <div className="stats-page__top-line">
          <AvatarHeading
            avatarHeadingText="Stats for Jack"
            headingStyle={"dropdown-heading"}
            className="stats-page__heading"
          />

          <div className="stats-page__buttons">
            <DropdownButton
              option1={"LEVEL 1"}
              option2={"LEVEL 2"}
              option3={"LEVEL 3"}
              option4={"LEVEL 4"}
              dropdownStyle={"dropdown-level"}
            />
            <DropdownButton
              option1={"JACK"}
              option2={"HOPE"}
              option3={"EMMA"}
              option4={"CONNOR"}
              dropdownStyle={"dropdown-name"}
            />
          </div>
        </div>

        <div className="stats-page__score-graph">
          <h2>Score Over Time</h2>
          <Graph />
        </div>

        <div className="stats-page__correct-phonemes">
          <h2>Correct Phonemes</h2>
          <ul>
            <li>SH</li>
            <li>EE</li>
            <li>P</li>
          </ul>
        </div>
      </section>
    </div>
  );
};

export default StatsPage;
