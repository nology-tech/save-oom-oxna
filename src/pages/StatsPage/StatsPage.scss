@use "../../styles/_variables.scss" as *;

.stats-page {
  background-color: $color-secondary-dark;

  &__top-line {
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin: 1rem;
    margin-left: 3rem;
  }

  &__heading {
    margin: 1rem;
  }

  &__buttons {
    width: 35%;
    display: flex;
    justify-content: space-evenly;
  }

  &__score-graph {
    margin-top: 1rem;
    margin-left: 3rem;
    padding: 1rem;
    width: 35rem;
    height: 18rem;
    background-color: $color-secondary-light;
  }

  &__correct-phonemes {
    margin-top: 1rem;
    margin-left: 3rem;
    margin-right: auto;
    margin-bottom: 2rem;
    padding: 1rem;
    width: 56rem;
    height: 13rem;
    background-color: $color-secondary-light;
  }

  &__buttons-name {
    background-color: black;
    border: 1px solid $color-accent;
  }
}

li {
  float: left;
  color: $color-white;
  list-style: none;
  margin: 0.5rem;
}

h2 {
  color: $color-white;
}

.stats-page__button-level {
  background-color: black;
  border: 1px solid $color-accent;
}
