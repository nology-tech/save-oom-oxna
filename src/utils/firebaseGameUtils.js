import { db } from "../firebase";
import {
  addDoc,
  collection,
  doc,
  getDocs,
  getDoc,
  query,
  where,
  orderBy,
  Timestamp,
} from "firebase/firestore";


/**
 * Gets a user object using the userId.
 * @param {*} userId 
 * @returns 
 */
export const getUserById = async( userId ) => {
  const docRef = doc(db, "users", userId);
  const docSnap = await getDoc( docRef );
  return docSnap;
}

/** 
 * Get all the rounds of a specific game for a given user, ordered by most recent first 
 @param {*} userId
*@returns a querySnap (document snapshot), returns a Promise.
*/

export const getGameRoundsForUser = async (id, game) => {
  const q = query(
    collection(db, "users", id, "rounds_played"),
    where("gameId", "==", game),
    orderBy("roundTime", "desc")
  );
  const querySnap = await getDocs(q);
  return querySnap;
};

/** 
   * Get all the correct round of a specific game for a given user, ordered by most recent first 
  @param {String} userId
  @param {String} gameId
  *@returns a promise with the query snapshot.
  */
export const getCorrectGameRoundsForUser = async (id, game) => {
  const q = query(
    collection(db, "users", id, "rounds_played"),
    where("gameId", "==", game),
    where("correct", "==", true),
    orderBy("roundTime", "desc")
  );
  const querySnap = await getDocs(q);
  return querySnap;
};

/** Get all the incorrect round of a specific game for a given user, ordered by most recent first 
  @param {String} userId
  @param {String} gameId
  *@returns  a promise with the query snapshot.
  */
export const getIncorrectGameRoundsForUser = async (id, game) => {
  const q = query(
    collection(db, "users", id, "rounds_played"),
    where("gameId", "==", game),
    where("correct", "==", false),
    orderBy("roundTime", "desc")
  );
  const querySnap = await getDocs(q);
  return querySnap;
};

/** Function that returns an array of rounds 
  @param {String} userId
  @param {String} gameId
  @param {Function} getGameRounds
  *@returns returns a promise.
  */
export const getArrayOfRounds = async (id, game, getGameRounds) => {
  const querySnap = await getGameRounds(id, game);
  const phonicsArr = querySnap.docs.map((doc) => {
    return doc.data().phonic;
  });
  return phonicsArr;
};

/** Function that returns an array of rounds 
  @param {String} id
  @param {String} game
  @param {Object} roundInfo
  @param {String} answer
  *@returns returs a promise.
  */
export const saveUserRound = async (id, game, level, gameStats, answer) => {
  const { isCorrect, score } = gameStats;
  await addDoc(collection(db, "users", id, "rounds_played"), {
    correct: isCorrect,
    gameId: game,
    phonic: answer,
    level: level,
    roundTime: Timestamp.fromDate(new Date()).toDate(),
    score: score,
  });
};
