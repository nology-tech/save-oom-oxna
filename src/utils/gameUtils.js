import phonicsData from "../data/phonicsData";
import { 
    // getGameRoundsForUser, 
    getCorrectGameRoundsForUser, 
    getIncorrectGameRoundsForUser,
    getArrayOfRounds
     } from "./firebaseGameUtils";

// const staticPhonicsArray = Object.keys(phonicsData.levelOne);

/**
 * Get a list of phonics from the static array, based on the level.
 * @param {*} level 
 * @returns 
 */
const getStaticPhonicsArray = (level) => {
    let levelData = null;
    switch(level) {
        case 1:
            levelData = "levelOne";
            break;
        case 2:
            levelData = "levelTwo";
            break;
        case 3:
            levelData = "levelThree";
            break;
        case 4:
            levelData = "levelFour";
            break;
        case 5:
            levelData = "levelFive";
            break;
    }
    return Object.keys( phonicsData[levelData] );
}

/**
 * Shuffle that array. Nice and simple algorithm.
 * Code courtesy of Geeks for Geeks.
 * @param {*} array 
 * @returns 
 */
const shuffleArray = (array) => {
    for (var i = array.length - 1; i > 0; i--) {
    
        // Generate random number
        var j = Math.floor(Math.random() * (i + 1));
                    
        var temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
        
    return array;
}
    

/**
 * Get the array of sounds for the swing game.
 * Attempts to return a mix of incorrect, correct and new sounds for the given child, based on their
 * past results. Basic algo at the moment!
 * @param {*} id 
 * @param {*} level 
 * @returns 
 */
export const getArrayForSwing = async (id, level) => {

    // 50 wrong = 8
    // 30 new = 4
    // fill up from rest = 3

    const MAX_PER_GAME = 15;
    const MAX_INCORRECT_PER_GAME = 8;
    const MAX_CORRECT_PER_GAME = 4;

    const game = "swing";

    const staticArray = getStaticPhonicsArray( level );
    console.log(`static array for level ${level}`, staticArray);

    const incorrect = await getArrayOfRounds( id, game, getIncorrectGameRoundsForUser);
    const correct = await getArrayOfRounds( id, game, getCorrectGameRoundsForUser);

    console.log( "correct", correct);
    console.log( "incorrect", incorrect);

    // all results
    const result = new Set();

    // add in at most MAX_INCORRECT_PER_GAME of incorrect answers
    for (let phonic of incorrect) {
        result.add(phonic);
        if( result.length >= MAX_INCORRECT_PER_GAME ) {
            console.log("Break!! Reach max incorrect");
            break;
        }
    }
    console.log("After adding incorrect", result);

    let maxCorrect = result.length + MAX_CORRECT_PER_GAME;
    // add in at most MAX_CORRECT_PER_GAME of correct answers
    for (let phonic of correct) {
        result.add(phonic);
        if( result.length >= maxCorrect ) {
            break;
        }
    }
    console.log("After adding correct", result);

    // fill up remaining
    for (let i = 0; i < staticArray.length; i++) {
        result.add( staticArray[i] );
        if( result.size >= MAX_PER_GAME ) {
            break;
        }
    }

    console.log("At end", result);

    const shuffled = shuffleArray( Array.from(result) );

    return shuffled;
}