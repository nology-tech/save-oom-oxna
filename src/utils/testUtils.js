import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import UserContext from "../contexts/UserContext";

export const customRender = (
  ui,
  useRouting = true,
  contextOptions = { useContext: true, value: undefined }
) => {
  // wrap components in routing and context if requested
  const routingUiResult = useRouting && wrapWithRouting(ui);
  const contextUiResult =
    contextOptions.useContext &&
    wrapWithContext(routingUiResult, contextOptions.value);

  // use RTL's render function to return the test component
  return render(contextUiResult);
};

const wrapWithRouting = (ui) => {
  return <Router>{ui}</Router>;
};

const wrapWithContext = (ui, value) => {
  return <UserContext.Provider value={value}>{ui}</UserContext.Provider>;
};
