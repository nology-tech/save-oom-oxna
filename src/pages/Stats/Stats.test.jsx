import { customRender } from "../../utils/testUtils";
import StatsPage from "./Stats";

it("Should render the Stats page", () => {
  const { container } = customRender(<StatsPage />);

  expect(container).toMatchSnapshot();
});
