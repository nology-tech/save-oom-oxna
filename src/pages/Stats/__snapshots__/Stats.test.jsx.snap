// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the Stats page 1`] = `
<div>
  <div
    class="layout"
  >
    <nav
      class="nav-bar"
    >
      <div
        class="nav-bar__section"
      >
        <a
          href="/dashboard"
        >
          <img
            alt="Logo image"
            class="logo__image"
            src="logo.png"
          />
        </a>
        <a
          href="/swing-game"
        >
          <div
            class="nav-bar-header"
          >
            <h2
              class="nav-bar-header__heading"
            >
              Swing Game
            </h2>
          </div>
        </a>
        <a
          href="/dashboard"
        >
          <div
            class="nav-bar-header"
          >
            <h2
              class="nav-bar-header__heading"
            >
              Dashboard
            </h2>
          </div>
        </a>
        <a
          href="/stats-page"
        >
          <div
            class="nav-bar-header"
          >
            <h2
              class="nav-bar-header__heading"
            >
              Stats
            </h2>
          </div>
        </a>
        <h2 />
      </div>
      <button
        class="button-logout"
      >
        Logout
      </button>
    </nav>
    <main
      class="layout__main"
    >
      <section
        class="stats-page"
      >
        <div
          class="stats-page__top-line"
        >
          <div
            class="avatar-heading"
          >
            <h1
              class="dropdown-heading"
            >
              Stats for Jack
            </h1>
          </div>
          <div
            class="stats-page__buttons"
          >
            <span>
              Dropdown #1
            </span>
            <span>
              Dropdown #2
            </span>
          </div>
        </div>
        <div
          class="stats-page__container"
        >
          <div
            class="stats-page__score-graph"
          >
            <h2>
              Score Over Time
            </h2>
            <div
              role="graph"
            />
          </div>
          <div
            class="stats-page__correct-phonemes"
          >
            <h2>
              Correct Phonemes
            </h2>
            <ul>
              <li>
                SH
              </li>
              <li>
                EE
              </li>
              <li>
                P
              </li>
            </ul>
          </div>
        </div>
      </section>
    </main>
  </div>
</div>
`;
